torch_xla/csrc/init_python_bindings.cpp

```diff
@@ -1373,72 +1373,76 @@ void InitXlaModuleBindings(py::module m) {
-            // the existing sharding first.
-            auto current_sharding_spec = xtensor->sharding_spec();
-            if (current_sharding_spec) {
-              XLA_CHECK(ShardingUtil::EqualShardingSpecs(
-                  *new_sharding_spec, *current_sharding_spec))
-                  << "Existing annotation must be cleared first.";
-              return;
-            }
+    // For data, we need to deal with the data transfers between
+    // host and device.
+    at::Tensor cpu_tensor;
+    if (xtensor->CurrentTensorData().has_value()) {
+      TORCH_LAZY_COUNTER("VirtualDeviceUsage", 1);
+      // When virtual device is enabled for SPMD, we defer the initial
+      // data transfer to the device and retain the original data on the
+      // host, until the sharded data transfer.
+      cpu_tensor = xtensor->CurrentTensorData().value();
+    } else {
+      // A new input tensor is not expected to be sharded. But sometimes,
+      // the same input is used sharding annotation, in which case we can
+      // skip if it's the same sharding; however, if it's the same input
+      // with a different sharding then we block & ask the user to clear
+      // the existing sharding first.
+      auto current_sharding_spec = xtensor->sharding_spec();
+      if (current_sharding_spec) {
+        XLA_CHECK(ShardingUtil::EqualShardingSpecs(*new_sharding_spec,
+                                                   *current_sharding_spec))
+            << "Existing annotation must be cleared first.";
+        return;
+      }
 
-            // If the at::Tensor data is not present, we need to re-download the
-            // tensor from the physical device to CPU. In that case, the value
-            // must be present on the backend device.
-            XLA_CHECK(xtensor->CurrentDataHandle() &&
-                      xtensor->CurrentDataHandle()->HasValue())
-                << "Cannot shard tensor. Data does not present on any device.";
-            std::vector<XLATensorPtr> xla_tensors{xtensor};
-            cpu_tensor = XLAGraphExecutor::Get()->GetTensors(&xla_tensors)[0];
-          }
-          auto xla_data = CreateTensorsData(
-              std::vector<at::Tensor>{cpu_tensor},
-              std::vector<XLATensor::ShardingSpecPtr>{new_sharding_spec},
-              std::vector<std::string>{GetVirtualDevice().toString()})[0];
-          xtensor->SetXlaData(xla_data);
-          xtensor->SetShardingSpec(*new_sharding_spec);
-
-          // Register sharded tensor data.
-          XLAGraphExecutor::Get()->RegisterTensor(xtensor->data());
-        });
+      // If the at::Tensor data is not present, we need to re-download the
+      // tensor from the physical device to CPU. In that case, the value
+      // must be present on the backend device.
+      XLA_CHECK((xtensor->CurrentDataHandle() &&
+                 xtensor->CurrentDataHandle()->HasValue()) ||
+                    device_data_node = nullptr)
```

Comment:
here is where I make check more generous, since DeviceData Node is equivalent to data handle.

Reply:
oh typo here.. should but `!=`

Reply:
I think I messed this up when fighting with formater.. let me rerun tests.